### PR TITLE
Revert #43546 "save subobligations in the projection cache"

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
This effectively reverts #43546 as it seems that it does affect performance more than the PR has anticipated.

This also removes the deduplication code from #48296 as duplications were primarily coming from #43546 and after removing that code it probably doesn't worth paying the cost of using a hash map.

Interestingly, the regression tests introduced in #43546 didn't fail after reverting, but some other tests had breakage. For now I just applied `--bless`, but I'd like to figure out if it could be a problem later on.

Preliminary perf results confirms that we no longer need deduplication to avoid exponential behavior.

image

This PR will be draft until I figure out the potential breakage.

r? @ghost

Need a perf run. A crater run will help as well to identify potential breakage.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/project.rs
Comment (jyn514) [inline compiler/rustc_trait_selection/src/traits/project.rs:549]: Any reason you removed the comment too?

Comment (ishitatsuyuki) [reply compiler/rustc_trait_selection/src/traits/project.rs:549]: I dropped it because I also removed the relevant code; that said, I think I deleted too much and some of them still serve as an explanation, so I will restore them later.

**Comment by ishitatsuyuki** [comment]
Perf summary: improvements on deeply-nested, around -0.5% for certain (probably assoc-type heavy) cases, and no change for vast of the benchmarks.

I'll continue to try to figure out what might break...

**Comment by ishitatsuyuki** [comment]
I've concluded that the changes in UI test stderr can be justified, as it's not a recursive obligation but just a (unresolvable) self-reference, and such cyclic definitions are unlikely in real world use cases to warrant a specific diagnostic. I did some additional cleanup to fully revert the related changes.

As for potential breakages, I feel it's unlikely to happen given all the tests are passing. Maybe we could get a crater run though.

r? @nikomatsakis since you authored the original changes

**Comment by nikomatsakis** [comment]
@ishitatsuyuki the primary motivation here is performance, or something else?

**Comment by ishitatsuyuki** [comment]
The primary motivation is not performance, probably.

I was looking at the `debug!` traces of the trait resolver in order to understand something else, but I end up discovering that seemingly simple predicates can resolve to a large number of obligations due to the duplicates, like the case below:

```
AssocTypeNormalizer: depth=0 normalized <T as num::dec2flt::rawfp::RawFloat>::Bits to <T as num::dec2flt::rawfp::RawFloat>::Bits, now with 20 obligations
```

So then I spent some time figuring out why does this happen, and the result is this PR. The motivation is not that strong, I just wanted to get rid of these redundant log entries.

**Comment by ishitatsuyuki** [comment]
Status: get_paranoid_cache_value_obligation is gone with #73905 and I've got a real test failure now, so I need some time to figure out a fix (or close this if I can't fix it).

**Comment by ishitatsuyuki** [comment]
New discovery: a side effect of this PR is that #76179 would be fixed. I don't know if this is the only way to fix it, though.

**Comment by ishitatsuyuki** [comment]
Turns out the (what I think are) regressions was actually caused by #76179, so I filed #78049. Not sure whether the fix of #76179 is correct or accidental.

**Comment by Dylan-DPC** [comment]
@nikomatsakis this is waiting for your review :)

**Comment by ishitatsuyuki** [comment]
Feel free to close this if it's too costly to review - I guess there's no practical motivation behind this and these kind of issues are better resolved with chalk.

**Comment by nikomatsakis** [comment]
Hi -- quick update. The last month or two, owing to Foundation work, has been really challenging for me to do reviews. I'm still somewhat uncertain about my take on this PR but I'm inclined to let it sit a bit longer and to give it one more read before we decide how to go forward. =)

**Comment by ishitatsuyuki** [comment]
Rebased. For some reason it seems that upstream changes have made the test changes unnecessary (probably due to additional checks).

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Comment by Dylan-DPC** [comment]
@nikomatsakis this is ready for review

**Comment by Mark-Simulacrum** [comment]
This will need a review from someone (more) familiar with the trait system, whether that's @nikomatsakis or perhaps @jackh726 or so. Not sure who exactly falls into the set here :)

**Comment by jyn514** [comment]
@matthewjasper might also be a good reviewer.

**Comment by jackh726** [comment]
I'm happy to take over the review here, but given that @nikomatsakis had (potential) reservations in https://github.com/rust-lang/rust/pull/77325#issuecomment-743983150, I'm going to see if I can try to go over this sometime this week with him. But I'll assign myself here to "push" this PR either to r+ or close.

r? @jackh726

**Comment by joshtriplett** [comment]
Given the amount of time that has passed since the previous perf run, I think it'd be useful to run perf again on the now-rebased version.

@bors try @rust-timer queue

**Comment by jyn514** [comment]
Those are *really* good numbers. I would love to see this merged.

**Review by lcnr — APPROVED** [review]
This looks good to me, though I also don't mind waiting for @nikomatsakis to look at it again.

With the nearing beta cutoff in #84909 I don't think there will be a lot of risk in merging this PR.

File: compiler/rustc_infer/src/traits/project.rs
Comment (ishitatsuyuki) [reply compiler/rustc_infer/src/traits/project.rs:165]: GitHub doesn't let me apply the change through the web UI, so I'll apply the change once I get back to my Linux box. Feel free to use the "maintainer edit" feature if it works and you need the change quickly.

**Comment by ishitatsuyuki** [comment]
Applied the nit and rebased.

**Comment by nikomatsakis** [comment]
So I was digging into the original issue and bringing it back in cache (pun intended). I've been somewhat concerned about landing this PR -- my concern is that the issue was very much tight to the exact execution order of everything. I'm wondering if the regression test has "bitrotted" but the original issue is still possible.

I'm trying to remember the details of how the bug manifested. It was something like this:

* We do the evaluation, which prompts us to do a normalization. This records in the cache a result which includes an inference variable, along with obligations that will cause that inference variable to be resolved.
* The evaluation completes but the cache entry remains. Those obligations are either not resolved or are discarded somehow.
* The cache entry is reused in another context and the inference variable never winds up being resolved.

**Comment by nikomatsakis** [comment]
I was going to come here and comment about the fact that #85382 intersected this PR, although not so deeply.

**Comment by nikomatsakis** [comment]
Maybe a good next step here would be a check only crater run? @ishitatsuyuki if you wanted to rebase, that'd be a good way to help try and flush out problems.

**Comment by ishitatsuyuki** [comment]
Rebased (strategy=ours). Can someone queue a check-only crater run per https://github.com/rust-lang/rust/pull/77325#issuecomment-845977862?

**Comment by nikomatsakis** [comment]
@bors try

I think this is the first step, right?

**Comment by Dylan-DPC** [comment]
i think we can run it directly with craterbot... 

@craterbot check

**Comment by ishitatsuyuki** [comment]
@Dylan-DPC try commit doesn't look correct...

> ⚠️ Try build based on commit 28bd098, but latest commit is a1815d7. Did you forget to make a new try build?

**Comment by Dylan-DPC** [comment]
ye just saw that will run it after the try build. thanks 

@craterbot cancel

**Comment by Dylan-DPC** [comment]
i think there's no way to cancel :D i'll retry-it once the try completes

**Comment by Dylan-DPC** [comment]
@craterbot run

**Comment by jackh726** [comment]
Looks like crater queue for this isn't correct. Both the older check and a newer "run" are queued. Should only be a new "check"

**Comment by Dylan-DPC** [comment]
@craterbot abort name=pr-77325

**Comment by Dylan-DPC** [comment]
@craterbot retry (just to ensure it's in the right state)

**Comment by Dylan-DPC** [comment]
@jackh726 removed the older one now :) thanks

**Comment by jackh726** [comment]
We don't need a test, just a check

@craterbot abort name=pr-77325-1

@craterbot check

**Comment by jackh726** [comment]
@craterbot check

**Comment by Dylan-DPC** [comment]
aah cool :)

**Comment by ishitatsuyuki** [comment]
It seems that this indeed introduces problems with trait resolution in some cases:

[Variant 1](https://crater-reports.s3.amazonaws.com/pr-77325/try%237c3a44cf5f95fa5cf164a312bdcf845157af7093/reg/capnp-rpc-0.14.1/log.txt)
[Variant 2](https://crater-reports.s3.amazonaws.com/pr-77325/try%237c3a44cf5f95fa5cf164a312bdcf845157af7093/reg/typ-0.1.1/log.txt)

Something similar happened when I updated the test cases, although all of them were compile-fail tests so I thought that fine and went with a bless. Turns out it isn't OK.

I probably will close this PR, unless there turns out to be an easy enough solution to tackle the regressions above.

**Comment by jackh726** [comment]
Maybe we should see if #84944 suffers from the same problems as this PR

**Pull request closed without merging**
